File: uikit/src/main/java/com/sendbird/uikit/widgets/RoundCornerView.java
package com.sendbird.uikit.widgets;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapShader;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.graphics.Shader;
import android.util.AttributeSet;
import android.view.ViewGroup;
import android.widget.ImageView;

import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatImageView;

import com.sendbird.uikit.R;

public class RoundCornerView extends ViewGroup {
    private final Paint roundingPaint = new Paint();
    private final RectF canvasBounds = new RectF();
    private final Canvas tempCanvas = new Canvas();
    private Bitmap tempBitmap;
    private float radius;
    private final AppCompatImageView child;

    public RoundCornerView(Context context) {
        this(context, null);
    }

    public RoundCornerView(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public RoundCornerView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        roundingPaint.setAntiAlias(true);
        roundingPaint.setFilterBitmap(true);

        this.radius = getContext().getResources().getDimension(R.dimen.sb_size_16);
        this.child = new AppCompatImageView(getContext());
        this.child.setScaleType(ImageView.ScaleType.CENTER_CROP);
        this.child.setMinimumWidth((int) getContext().getResources().getDimension(R.dimen.sb_size_100));
        this.child.setMinimumHeight((int) getContext().getResources().getDimension(R.dimen.sb_size_100));
        this.child.setMaxWidth((int) getContext().getResources().getDimension(R.dimen.sb_message_max_width));
        this.child.setMaxHeight((int) getContext().getResources().getDimension(R.dimen.sb_message_max_height));
        addView(child);
    }

    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {
        child.layout(0, 0, getMeasuredWidth(), getMeasuredHeight());
    }

    public void setRadius(float radius) {
        this.radius = radius;
    }

    @Override
    protected void dispatchDraw(Canvas canvas) {
        int width = getWidth();
        int height = getHeight();
        if (isInEditMode()) {
            Paint debugPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
            debugPaint.setColor(getResources().getColor(R.color.background_400));
            canvas.drawRoundRect(new RectF(0, 0, width, height), 0, 0, debugPaint);
            return;
        }

        if (width <= 0 || height <=0) {
            super.dispatchDraw(canvas);
            return;
        }

        if (tempBitmap == null || tempBitmap.isRecycled() || (tempBitmap.getWidth() != getWidth() || tempBitmap.getHeight() != getHeight())) {
            if (tempBitmap != null) {
                tempBitmap.recycle();
            }
            try {
                tempBitmap = Bitmap.createBitmap(getWidth(), getHeight(), Bitmap.Config.ARGB_8888);
            } catch (OutOfMemoryError e) {
                tempBitmap = Bitmap.createBitmap(getWidth(), getHeight(), Bitmap.Config.RGB_565);
            }
        } else {
            tempBitmap.eraseColor(Color.TRANSPARENT);
        }

        tempCanvas.setBitmap(tempBitmap);
        super.dispatchDraw(tempCanvas); // draw children
        roundingPaint.setShader(new BitmapShader(tempBitmap, Shader.TileMode.CLAMP, Shader.TileMode.CLAMP));
        canvasBounds.set(0, 0, width, height);
        canvas.drawRoundRect(canvasBounds, radius, radius, roundingPaint);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        int specWidth = MeasureSpec.getSize(widthMeasureSpec);
        int specHeight = MeasureSpec.getSize(heightMeasureSpec);
        child.measure(MeasureSpec.makeMeasureSpec(specWidth, MeasureSpec.EXACTLY), MeasureSpec.makeMeasureSpec(specHeight, MeasureSpec.EXACTLY));
        setMeasuredDimension(child.getMeasuredWidth(), child.getMeasuredHeight());
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();

        if (tempBitmap != null) {
            tempBitmap.recycle();
        }
    }

    public ImageView getContent() {
        return child;
    }
}

File: uikit/src/main/java/com/sendbird/uikit/internal/ui/widgets/RoundCornerView.kt
package com.sendbird.uikit.internal.ui.widgets

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapShader
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import android.graphics.RectF
import android.graphics.Shader
import android.util.AttributeSet
import android.view.ViewGroup
import android.widget.ImageView
import androidx.appcompat.widget.AppCompatImageView
import androidx.core.content.ContextCompat
import com.sendbird.uikit.R

internal class RoundCornerView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : ViewGroup(context, attrs, defStyleAttr) {
    private val roundingPaint = Paint()
    private val canvasBounds = RectF()
    private val tempCanvas = Canvas()
    private var tempBitmap: Bitmap? = null
    private val child: AppCompatImageView
    var radius: Float
    var cornerRadii: FloatArray? = null
    val content: ImageView
        get() = child

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
        child.layout(0, 0, measuredWidth, measuredHeight)
    }

    override fun dispatchDraw(canvas: Canvas) {
        val width = width
        val height = height
        if (isInEditMode) {
            val debugPaint = Paint(Paint.ANTI_ALIAS_FLAG)
            debugPaint.color = ContextCompat.getColor(context, R.color.background_400)
            canvas.drawRoundRect(RectF(0f, 0f, width.toFloat(), height.toFloat()), 0f, 0f, debugPaint)
            return
        }
        if (width <= 0 || height <= 0) {
            super.dispatchDraw(canvas)
            return
        }
        tempBitmap?.run {
            if (isRecycled || width != getWidth() || height != getHeight()) {
                tempBitmap?.recycle()
            } else {
                eraseColor(Color.TRANSPARENT)
            }
            tempBitmap = try {
                Bitmap.createBitmap(getWidth(), getHeight(), Bitmap.Config.ARGB_8888)
            } catch (e: OutOfMemoryError) {
                Bitmap.createBitmap(getWidth(), getHeight(), Bitmap.Config.RGB_565)
            }
        } ?: run {
            tempBitmap = try {
                Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
            } catch (e: OutOfMemoryError) {
                Bitmap.createBitmap(width, height, Bitmap.Config.RGB_565)
            }
        }
        tempCanvas.setBitmap(tempBitmap)
        super.dispatchDraw(tempCanvas) // draw children
        roundingPaint.shader = BitmapShader(tempBitmap!!, Shader.TileMode.CLAMP, Shader.TileMode.CLAMP)
        canvasBounds[0f, 0f, width.toFloat()] = height.toFloat()
        cornerRadii?.let {
            val path = Path()
            path.addRoundRect(canvasBounds, it, Path.Direction.CW)
            canvas.drawPath(path, roundingPaint)
        } ?: run {
            canvas.drawRoundRect(canvasBounds, radius, radius, roundingPaint)
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val specWidth = MeasureSpec.getSize(widthMeasureSpec)
        val specHeight = MeasureSpec.getSize(heightMeasureSpec)
        child.measure(
            MeasureSpec.makeMeasureSpec(specWidth, MeasureSpec.EXACTLY),
            MeasureSpec.makeMeasureSpec(specHeight, MeasureSpec.EXACTLY)
        )
        setMeasuredDimension(child.measuredWidth, child.measuredHeight)
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        tempBitmap?.recycle()
    }

    init {
        roundingPaint.isAntiAlias = true
        roundingPaint.isFilterBitmap = true
        radius = getContext().resources.getDimension(R.dimen.sb_size_16)
        child = AppCompatImageView(getContext()).apply {
            scaleType = ImageView.ScaleType.CENTER_CROP
            minimumWidth = getContext().resources.getDimension(R.dimen.sb_size_100).toInt()
            minimumHeight = getContext().resources.getDimension(R.dimen.sb_size_100).toInt()
            maxWidth = getContext().resources.getDimension(R.dimen.sb_message_max_width).toInt()
            maxHeight = getContext().resources.getDimension(R.dimen.sb_message_max_height).toInt()
        }
        addView(child)
    }
}
